node/src/client.rs

```diff
@@ -0,0 +1,269 @@
+use crate::{
+	runtime::HostRuntimeApis,
+	service::{ComposableExecutor, PicassoExecutor},
+};
+pub use common::{AccountId, Balance, BlockNumber, Hash, Header, Index, OpaqueBlock as Block};
+use sc_client_api::{Backend as BackendT, BlockchainEvents, KeyIterator};
+use sc_executor::NativeElseWasmExecutor;
+use sc_service::{TFullBackend, TFullClient};
+use sp_api::{CallApiAt, NumberFor, ProvideRuntimeApi};
+use sp_blockchain::HeaderBackend;
+use sp_consensus::BlockStatus;
+use sp_runtime::{
+	generic::{BlockId, SignedBlock},
+	traits::{BlakeTwo256, Block as BlockT},
+	Justifications,
+};
+use sp_storage::{ChildInfo, PrefixedStorageKey, StorageData, StorageKey};
+use std::sync::Arc;
+
+pub type FullClient<RuntimeApi, Executor> =
+	TFullClient<Block, RuntimeApi, NativeElseWasmExecutor<Executor>>;
+pub type FullBackend = TFullBackend<Block>;
+type PicassoClient = FullClient<picasso_runtime::RuntimeApi, PicassoExecutor>;
+type ComposableClient = FullClient<composable_runtime::RuntimeApi, ComposableExecutor>;
+
+/// Config that abstracts over all available client implementations.
+///
+/// For a concrete type there exists [`Client`].
+pub trait AbstractClient<Block, Backend>:
+	BlockchainEvents<Block>
+	+ Sized
+	+ Send
+	+ Sync
+	+ ProvideRuntimeApi<Block>
+	+ HeaderBackend<Block>
+	+ CallApiAt<Block, StateBackend = Backend::State>
+where
+	Block: BlockT,
+	Backend: BackendT<Block>,
+	Backend::State: sp_api::StateBackend<BlakeTwo256>,
+	Self::Api: HostRuntimeApis<StateBackend = Backend::State>,
+{
+}
+
+impl<Block, Backend, Client> AbstractClient<Block, Backend> for Client
+where
+	Block: BlockT,
+	Backend: BackendT<Block>,
+	Backend::State: sp_api::StateBackend<BlakeTwo256>,
+	Client: BlockchainEvents<Block>
+		+ ProvideRuntimeApi<Block>
+		+ HeaderBackend<Block>
+		+ Sized
+		+ Send
+		+ Sync
+		+ CallApiAt<Block, StateBackend = Backend::State>,
+	Client::Api: HostRuntimeApis<StateBackend = Backend::State>,
+{
+}
+
+/// A client instance of Picasso.
+#[derive(Clone)]
+pub enum Client {
+	/// Picasso client type
+	Picasso(Arc<PicassoClient>),
+	/// Composable client type
+	Composable(Arc<ComposableClient>),
+}
+
+impl From<Arc<PicassoClient>> for Client {
+	fn from(client: Arc<PicassoClient>) -> Self {
+		Self::Picasso(client)
+	}
+}
+
+impl From<Arc<ComposableClient>> for Client {
+	fn from(client: Arc<ComposableClient>) -> Self {
+		Self::Composable(client)
+	}
+}
+
+macro_rules! match_client {
```

Comment:
there is crate which does this https://docs.rs/enum_dispatch/latest/enum_dispatch/ kind of

Reply:
can only use it if our crate defines the trait.